File: python/sirius_python/src/pyring/elements.py
import passmethods

PassMethods = passmethods.PassMethods

class marker(object):
    def __init__(self,     
         fam_name, 
         length      = 0,
         nr_steps    = 1,
         pass_method = PassMethods.pm_identity_pass,
         angle       = None,
         gap         = None,
         polynom_a   = None,
         polynom_b   = None,
         k           = None,
         kl          = None,
         s           = None,
         sl          = None,
         kick_angle  = None,
         frequency   = None,
         voltage     = None,
         energy      = None,
         t_in        = None,
         r_in        = None,
         angle_in    = None,
         fint_in     = None,
         t_out       = None,
         r_out       = None,
         angle_out   = None,
         fint_out    = None,
         **kwargs
         ):
        
        self.fam_name    = fam_name
        self.length      = length
        self.nr_steps    = nr_steps
        self.pass_method = pass_method
        
        if angle      is not None: self.angle       = angle
        if gap        is not None: self.gap         = gap
        if polynom_a  is not None: self.polynom_a   = polynom_a
        if polynom_b  is not None: self.polynom_b   = polynom_b
        
        if k          is not None: self.k           = k
        if kl         is not None: self.kl          = kl
        if sl         is not None: self.sl          = sl
        if kick_angle is not None: self.kick_angle  = kick_angle
        if voltage    is not None: self.voltage     = voltage
        if frequency  is not None: self.frequency   = frequency
        if energy     is not None: self.energy      = energy
        
        if t_in       is not None: self.t_in        = t_in
        if r_in       is not None: self.r_in        = r_in
        if angle_in   is not None: self.angle_in    = angle_in
        if fint_in    is not None: self.fint_in     = fint_in
        
        if t_out      is not None: self.t_out       = t_out
        if r_out      is not None: self.r_out       = r_out
        if angle_out  is not None: self.angle_out   = angle_out 
        if fint_out   is not None: self.fint_out    = fint_out
        
    def __str__(self):
        r  = ''
        try: r += '   FamName: ' + self.fam_name + '\n' 
        except: pass
        try: r += '    Length: ' + str(self.length) + '\n'
        except: pass
        try: r += 'PassMethod: ' + str(self.pass_method) + '\n'
        except: pass
        try: r += '     Angle: ' + str(self.angle) + '\n'
        except: pass
        try: r += '   AngleIn: ' + str(self.angle_in) + '\n'
        except: pass
        try: r += '  AngleOut: ' + str(self.angle_out) + '\n'
        except: pass
        try: r += '       Gap: ' + str(self.gap) + '\n'
        except: pass
        try: r += '  PolynomA: ' + str(self.polynom_a) + '\n'
        except: pass
        try: r += '  PolynomB: ' + str(self.polynom_b) + '\n'
        except: pass
        try: r += '         K: ' + str(self.k) + '\n'
        except: pass
        try: r += '        KL: ' + str(self.kl) + '\n'
        except: pass
        try: r += '        SL: ' + str(self.sl) + '\n'
        except: pass
        try: r += ' KickAngle: ' + str(self.kick_angle) + '\n'
        except: pass
        try: r += '   Voltage: ' + str(self.voltage) + '\n'
        except: pass
        try: r += ' Frequency: ' + str(self.frequency) + '\n'
        except: pass
        try: r += '    Energy: ' + str(self.energy) + '\n'
        except: pass
        try: r += '       TIn: ' + str(self.t_in) + '\n'
        except: pass
        try: r += '       RIn: ' + str(self.r_in) + '\n'
        except: pass
        try: r += '    FIntIn: ' + str(self.fint_in) + '\n'
        except: pass
        try: r += '      TOut: ' + str(self.t_out) + '\n'
        except: pass
        try: r += '      ROut: ' + str(self.r_out) + '\n'
        except: pass
        try: r += '   FIntOut: ' + str(self.fint_out) + '\n'
        except: pass
        return r    
    
    
class drift(marker):
    def __init__(self, pass_method = PassMethods.pm_drift_pass, **kwargs):
        marker.__init__(self, pass_method = pass_method, **kwargs)

class corrector(marker):
    def __init__(self, pass_method = PassMethods.pm_corrector_pass, kick_angle  = None, **kwargs):
        if kick_angle == None: kick_angle = [0,0]
        marker.__init__(self, pass_method = pass_method, kick_angle = kick_angle, **kwargs)
        
class rfcavity(marker):
    def __init__(self, voltage, frequency, energy, pass_method = PassMethods.pm_cavity_pass, **kwargs):
        marker.__init__(self, pass_method = pass_method, voltage = voltage, frequency = frequency, energy = energy, **kwargs)

class quadrupole(marker):
    def __init__(self, pass_method = PassMethods.pm_str_mpole_symplectic4_pass, nr_steps = 10, polynom_a = None, polynom_b = None, **kwargs): 
        if polynom_a == None: polynom_a = [0,0]
        if polynom_b == None: polynom_b = [0,0] 
        if 'k' in kwargs: 
            if ('kl' in kwargs) or (pass_method == PassMethods.pm_thinquad_pass):
                raise Exception('Inconsistent marker parameters!')
            polynom_b[1] = kwargs['k']
        if 'kl' in kwargs:
            if ('k' in kwargs) or (pass_method != PassMethods.pm_thinquad_pass):
                raise Exception('Inconsistent marker parameters!')
        marker.__init__(self, pass_method = pass_method, nr_steps = nr_steps, polynom_a = polynom_a, polynom_b = polynom_b, **kwargs)

class sextupole(marker):
    def __init__(self, pass_method = 'str_mpole_symplectic4_pass', nr_steps = 5, polynom_a = None, polynom_b = None, **kwargs): 
        if polynom_a == None: polynom_a = [0,0,0]
        if polynom_b == None: polynom_b = [0,0,0] 
        if 's' in kwargs: 
            if ('sl' in kwargs) or (pass_method == 'thinsext_pass'):
                raise Exception('Inconsistent marker parameters!')
            pass_method  = 'str_mpole_symplectic4_pass'
            polynom_b[2] = kwargs['s']
        if 'sl' in kwargs:
            if ('s' in kwargs) or (pass_method != 'thinsext_pass'):
                raise Exception('Inconsistent marker parameters!')
        marker.__init__(self, pass_method = pass_method, nr_steps = nr_steps, polynom_a = polynom_a, polynom_b = polynom_b, **kwargs)

class rbend(marker):
    def __init__(self, length, angle, pass_method = 'bnd_mpole_symplectic4_pass', nr_steps = 10, polynom_a = None, polynom_b = None, gap = 0, **kwargs):
        if polynom_a == None: polynom_a = [0,0]
        if polynom_b == None: polynom_b = [0,0]
        if 'k' in kwargs:
            polynom_b[1] = kwargs['k']
        if 'angle_in' not in kwargs:
            kwargs['angle_in'] = 0.5 * angle
        if 'angle_out' not in kwargs:
            kwargs['angle_out'] = 0.5 * angle
        marker.__init__(self, pass_method = pass_method, length = length, angle = angle, gap = gap, nr_steps = nr_steps, polynom_a = polynom_a, polynom_b = polynom_b, **kwargs)

File: trackc++/elements.cpp
// TRACKC++
// ========
// Author: 		Ximenes R. Resende
// email:  		xresende@gmail.com, ximenes.resende@lnls.br
// affiliation:	LNLS - Laboratorio Nacional de Luz Sincrotron
// Date: 		Tue Dec 10 17:57:20 BRST 2013


#include "auxiliary.h"
#include "elements.h"
#include <cfloat>

const std::vector<double> Element::default_polynom = std::vector<double>(3,0);


// default constructor (constructs a drift element)
Element::Element(const std::string& fam_name_, const double& length_) :
    fam_name(fam_name_), length(length_) {
	for(unsigned int i=0; i<6; i++) {
		t_in[i] = t_out[i] = 0.0;
		for(unsigned int j=0; j<6; ++j) {
			if (i == j) {
				r_in[i*6+j] = r_out[i*6+j] = 1.0;
			} else {
				r_in[i*6+j] = r_out[i*6+j] = 0.0;
			}
		}
	}

};

Element Element::marker (const std::string& fam_name_) {
	Element e = Element(fam_name_, 0);
	e.pass_method = PassMethod::pm_identity_pass;
	return e;
}

Element Element::drift (const std::string& fam_name_, const double& length_) {
	Element e = Element(fam_name_, length_);
	e.pass_method = PassMethod::pm_drift_pass;
	return e;
}

Element Element::hcorrector(const std::string& fam_name_, const double& length_, const double& hkick_) {
	Element e = Element(fam_name_, length_);
	e.pass_method = PassMethod::pm_corrector_pass;
	e.hkick = hkick_;
	return e;
}

Element Element::vcorrector(const std::string& fam_name_, const double& length_, const double& vkick_) {
	Element e = Element(fam_name_, length_);
	e.pass_method = PassMethod::pm_corrector_pass;
	e.vkick = vkick_;
	return e;
}

Element Element::corrector(const std::string& fam_name_, const double& length_, const double& hkick_, const double& vkick_) {
	Element e = Element(fam_name_, length_);
	e.pass_method = PassMethod::pm_corrector_pass;
	e.hkick = hkick_;
	e.vkick = vkick_;
	return e;
}


Element Element::quadrupole (const std::string& fam_name_, const double& length_, const double& K_, const int nr_steps_) {
	Element e = Element(fam_name_, length_);
	e.pass_method = PassMethod::pm_str_mpole_symplectic4_pass;
	e.polynom_b[1] = K_;
	return e;
}

Element Element::sextupole (const std::string& fam_name_, const double& length_, const double& S_, const int nr_steps_) {
	Element e = Element(fam_name_, length_);
	e.nr_steps = nr_steps_;
	e.pass_method = PassMethod::pm_str_mpole_symplectic4_pass;
	e.polynom_b[2] = S_;
	return e;
}
Element Element::rbend (const std::string& fam_name_, const double& length_, const double& angle_, const double& angle_in_, const double& angle_out_, const double& K_, const double& S_) {
	Element e = Element(fam_name_, length_);
	e.pass_method = PassMethod::pm_bnd_mpole_symplectic4_pass;
	e.angle = angle_;
	e.angle_in = angle_in_;
	e.angle_out = angle_out_;
	e.polynom_b[1] = K_;
	e.polynom_b[2] = S_;
	return e;
}

Element Element::rfcavity (const std::string& fam_name_, const double& length_, const double& frequency_, const double& voltage_) {
	Element e = Element(fam_name_, length_);
	e.pass_method = PassMethod::pm_cavity_pass;
	e.frequency = frequency_;
	e.voltage = voltage_;
	return e;
}



void print_polynom(std::ostream& out, const std::string& label, const std::vector<double>& polynom) {
	int order = 0;
	for(unsigned int i=0; i<polynom.size(); ++i) {
		if (polynom[i] != 0) order = i+1;
	}
	if (order > 0) out << label;
	for(int i=0; i<order; ++i) {
		out << polynom[i] << " ";
	}
	if (order > 0) out << std::endl;
}

std::ostream& operator<< (std::ostream &out, const Element& el) {

	                      out << "fam_name      : " << el.fam_name << std::endl;
	if (el.length != 0)   out << "length        : " << el.length << std::endl;
	                      out << "pass_method   : " << pm_dict[el.pass_method] << std::endl;
	if (el.nr_steps > 1)  out << "nr_steps      : " << el.nr_steps << std::endl;
	if (el.thin_KL != 0)  out << "thin_KL       : " << el.thin_KL << std::endl;
	if (el.thin_SL != 0)  out << "thin_SL       : " << el.thin_SL << std::endl;
	if (el.angle != 0)    out << "bending_angle : " << el.angle << std::endl;
	if (el.angle != 0)    out << "entrance_angle: " << el.angle_in << std::endl;
	if (el.angle != 0)    out << "exit_angle    : " << el.angle_out << std::endl;
	if ((el.gap != 0) and ((el.fint_in != 0) or (el.fint_out != 0))) {
		                  out << "gap           : " << el.gap << std::endl;
		                  out << "fint_in       : " << el.fint_in << std::endl;
		                  out << "fint_out      : " << el.fint_out << std::endl;
	}
	print_polynom(        out,   "polynom_a     : ", el.polynom_a);
	print_polynom(        out,   "polynom_b     : ", el.polynom_b);
	if (el.frequency != 0)out << "frequency     : " << el.frequency << std::endl;
	if (el.voltage != 0)  out << "voltage       : " << el.voltage << std::endl;
	return out;
}



